### PR TITLE
Learn from calls to Equals methods in NullableWalker

### LINKED ISSUE #36591: Nullable analysis should learn from calls to well-known Equals methods
From @jcouv:

Update the null-state at the end of an invocation of a recognized method
https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-05-15.md#equality-postconditions

Comparison with null literal (or null constant): `object.Equals(x, null)`
- learn that `x` is null in when-true branch
- learn that `x` is not-null in when-false branch

Comparison with a not-null expression: `object.Equals(x, notNull)`
- learn that `x` is not-null in when-true branch
- for `object.Equals(x?.y, notNull)`, learn that x was not-null (the same as `x?.y == notNull`)

`object.Equals(x, maybeNull)`
- learn nothing about x

Open question: what are all the methods we recognize? So far:

- `Object.ReferenceEquals`
- `Object.Equals` 
- `IEqualityComparer<T>.Equals` 
- `EqualityComparer<T>.Equals` 
- `IEquatable<T>.Equals`

Also add some tests around user-defined `==` (we learn the same thing regardless of if it's builtin or user-defined)

### PR DESCRIPTION
**Pull request by RikkiGibson**
Resolves #36591

This changes the NullableWalker so that we can learn from calls to:
-  `static bool object.Equals(object, object)`,
- `static bool object.ReferenceEquals(object, object)`
- `bool object.Equals(object)` and overrides
- `bool IEquatable<T>(T)` and implementations
- `bool IEqualityComparer<T>(T, T)` and implementations

Scenarios where we can learn new things about null state:

```cs
object.[Reference]Equals(maybeNull, notNull) // when true, changes maybeNull to not-null
object.[Reference]Equals(notNull, null) // when true, changes notNull to maybe-null
object.[Reference]Equals(maybeNull, null) // when false, changes notNull to maybe-null
notNull.Equals(maybeNull) // when true, changes maybeNull to not-null
notNull.Equals(maybeNull) // when true, changes maybeNull to not-null
notNull.Equals(null) // when true, changes nonNull to maybe-null
```

The `notNull.Equals(null)` scenario is a little silly (if you called Equals on a null, didn't you already throw?) But I think that this is just such an antipattern that it doesn't matter if we don't do the perfect thing here.

I tried to think of situations where we could get in trouble when using object.ReferenceEquals on nullable value types, but I couldn't think of any where we incorrectly rule out a diagnostic in a scenario that throws an exception at runtime.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -109527,5 +109527,764 @@ class Program
             var comp = CreateCompilation(source);
             comp.VerifyDiagnostics();
         }
+
+        [Fact, WorkItem(35816, "https://github.com/dotnet/roslyn/issues/35816")]
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >Fact [](start = 9, length = 4)

I'll review tomorrow, but one quick note: please update the speclet: [https://github.com/dotnet/roslyn/blob/master/docs/features/nullable-reference-types.md#null-tests](https://github.com/dotnet/roslyn/blob/master/docs/features/nullable-reference-types.md#null-tests)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -2825,6 +2838,133 @@ private void ReinferMethodAndVisitArguments(BoundCall node, TypeWithState receiv
             SetResult(node, GetReturnTypeWithState(method), method.ReturnTypeWithAnnotations);
         }
 
+        private bool IsStaticEqualsMethod(MethodSymbol method)
+        {
+            if (method.Equals(compilation.GetSpecialTypeMember(SpecialMember.System_Object__EqualsObjectObject)))
+            {
+                return true;
+            }
+            else if (method.Equals(compilation.GetSpecialTypeMember(SpecialMember.System_Object__ReferenceEquals)))
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >if (method.Equals(compilation.GetSpecialTypeMember(SpecialMember.System_Object__ReferenceEquals))) [](start = 17, length = 98)

nit: `else if` can be simplified to `else { return method.Equals(compilation.GetSpecialTypeMember(SpecialMember.System_Object__ReferenceEquals)); } `

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -2813,8 +2813,21 @@ private void ReinferMethodAndVisitArguments(BoundCall node, TypeWithState receiv
                 method = (MethodSymbol)AsMemberOfType(receiverType.Type, method);
             }
 
-            method = VisitArguments(node, node.Arguments, refKindsOpt, method.Parameters, node.ArgsToParamsOpt,
-                node.Expanded, node.InvokedAsExtensionMethod, method).method;
+            ImmutableArray<VisitArgumentResult> results;
+            (method, results) = VisitArguments(node, node.Arguments, refKindsOpt, method.Parameters, node.ArgsToParamsOpt,
+                node.Expanded, node.InvokedAsExtensionMethod, method);
+
+            var arguments = node.Arguments;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >var [](start = 12, length = 3)

Consider moving all the new logic (the following block) into a new method and turning the single-use helper methods (like `IsStaticEqualsMethod`) into local functions.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 4]
```diff
@@ -2825,6 +2838,133 @@ private void ReinferMethodAndVisitArguments(BoundCall node, TypeWithState receiv
...
+            {
+                return false;
+            }
+
+            var parameterType = method.Parameters[0].TypeWithAnnotations;
+            var constructedIEquatableType = iEquatableType.Construct(ImmutableArray.Create(parameterType));
+            var members = constructedIEquatableType.GetMembers(WellKnownMemberNames.ObjectEquals);
+            if (members.Length != 1)
+            {
+                // If the Equals method is missing on the interface definition, or it's overloaded, give up.
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >or it's overloaded, give up. [](start = 80, length = 28)

I don't imagine that the BCL is going to add an `Equals` overload on `object`, but it's technically possible for one to be there. Can we pick the one with the `object` parameter type?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 5]
```diff
@@ -2825,6 +2838,133 @@ private void ReinferMethodAndVisitArguments(BoundCall node, TypeWithState receiv
...
+                return false;
+            }
+
+            var constructedIEquatableEqualsMethod = members[0];
+            if (constructedIEquatableEqualsMethod.Equals(method))
+            {
+                return true;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >return true; [](start = 16, length = 12)

This method has two code paths that could return true. I didn't understand why we need that. It seems like the path that uses `FindImplementationForInterfaceMember` would always work, no?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 6]
```diff
@@ -2825,6 +2838,133 @@ private void ReinferMethodAndVisitArguments(BoundCall node, TypeWithState receiv
...
+            if (constructedIEqualityComparerEqualsMethod.Equals(method))
+            {
+                return true;
+            }
+
+            var implementationMethod = method.ContainingType.FindImplementationForInterfaceMember(constructedIEqualityComparerEqualsMethod);
+            return method.Equals(implementationMethod);
+        }
+
+        private void LearnFromEqualsMethod(BoundExpression left, TypeWithState leftType, BoundExpression right, TypeWithState rightType)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >LearnFromEqualsMethod [](start = 21, length = 21)

Probably deserves a comment for documentation:
- a `null` argument tells us the other argument is maybe-null (ie. this is a pure null test)
- equality to an expression that is not-null means the argument is not-null

I see that the call to `LearnFromNullTest` in `AfterLeftChildHasBeenVisited` is preceded by a call to `SkipReferenceConversions`. I wonder if we need to do the same here. It would be good to test `IEquatable<B>.Equals(a, null)` where `a` has implicit reference conversion to type `B`. Testing with user-defined conversions which flip nullability would be useful too. 
We may have an issue tracking a problem in this area already.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: 1. Should we expect `InstanceObjectEqualsMethod_MaybeNullExprReferenceConversion_NonNullExpr_NoWarningWhenTrue` to catch this scenario?
2. LearnFromNullTest also strips conversions internally (calls RemoveConversion on the expression ) so I'm also wondering if both steps are even necessary.

---
In reply to: [297388081](https://github.com/dotnet/roslyn/pull/36722#discussion_r297388081) [](ancestors = 297388081)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 7]
```diff
@@ -109527,5 +109527,764 @@ class Program
...
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                    // (12,17): warning CS8602: Dereference of a possibly null reference.
+                    //             _ = c1.ToString(); // 1
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1").WithLocation(12, 17),
+                    // (16,17): warning CS8602: Dereference of a possibly null reference.
+                    //             _ = c1.ToString(); // 2
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c1").WithLocation(16, 17));
+        }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >} [](start = 8, length = 1)

Consider adding tests that don't rely on `string?` or `C? types, but rather also try `int?` or `TOpen?`.
Also, let's adding some tests for `notNull1` compared to `notNull2` (no learning, both remain not-null) and `maybeNull1` compared to `maybeNull2` (no learning, both remain maybe-null)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 8]
```diff
@@ -2825,6 +2838,133 @@ private void ReinferMethodAndVisitArguments(BoundCall node, TypeWithState receiv
...
+                return false;
+            }
+
+            var parameterType = method.Parameters[0].TypeWithAnnotations;
+            var constructedIEqualityComparerType = iEqualityComparerType.Construct(ImmutableArray.Create(parameterType));
+            var members = constructedIEqualityComparerType.GetMembers(WellKnownMemberNames.ObjectEquals);
+            if (members.Length != 1)
+            {
+                // If the Equals method is missing on the interface definition, or it's overloaded, give up.
+                return false;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >return false; [](start = 16, length = 13)

do we have tests covering all the branches of these methods (such as missing or malformed types/members)?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 4 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: The more "common" scenario is that `bool IEquatable<T>.Equals(T)` is not defined at all, but this is really just a "degrade gracefully in the face of bad metadata" behavior.

re: `GetWellKnownTypeMember`, I would like to, however I am stumped as to how to actually make the comparison between the method inside the original definition (`IEquatable<T>.Equals(T)`) and the method inside a constructed symbol (`IEquatable<string?>.Equals(string?)`)

---
In reply to: [297381621](https://github.com/dotnet/roslyn/pull/36722#discussion_r297381621) [](ancestors = 297381621)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 6 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Yes, `InstanceObjectEqualsMethod_MaybeNullExprReferenceConversion_NonNullExpr_NoWarningWhenTrue` is good.
I think we can probably remove `SkipReferenceConversions` from `AfterLeftChildHasBeenVisited` then. If a test is affected, then it can tell us what test to add for learning from Equals method.

---
In reply to: [297389419](https://github.com/dotnet/roslyn/pull/36722#discussion_r297389419) [](ancestors = 297389419,297388081)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 5 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: I found `m1.FindImplementationForInterfaceMember(m2)` returned null when m1 was equal to m2. I can double check on it.

---
In reply to: [297383790](https://github.com/dotnet/roslyn/pull/36722#discussion_r297383790) [](ancestors = 297383790)

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: I double checked on this, removing this return breaks the following tests
- IEqualityComparerEqualsMethod_MaybeNullExpr_NonNullExpr_NoWarningWhenTrue
- IEqualityComparer_SubInterfaceEqualsMethod_MaybeNullExpr_NonNullExpr_NoWarningWhenTrue
- IEqualityComparer_SubSubInterfaceEqualsMethod_MaybeNullExpr_NonNullExpr_NoWarningWhenTrue
- IEquatableEqualsMethod_MaybeNullExpr_NonNullExpr_NoWarningWhenTrue

---
In reply to: [297392516](https://github.com/dotnet/roslyn/pull/36722#discussion_r297392516) [](ancestors = 297392516,297383790)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 9]
```diff
@@ -2164,8 +2164,6 @@ protected override void AfterLeftChildHasBeenVisited(BoundBinaryOperator binary)
 
                 if (operandComparedToNull != null)
                 {
-                    operandComparedToNull = SkipReferenceConversions(operandComparedToNull);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >operandComparedToNull = SkipReferenceConversions(operandComparedToNull); [](start = 20, length = 72)

FYI @jcouv I found removing this conversion-stripping step did not affect anything.

**Comment by RikkiGibson** [comment]
@333fred I resolved some conflicts with your PR for ShouldRunNullableDirective. If you have a moment please take a glance at [86084e1](https://github.com/dotnet/roslyn/pull/36722/commits/86084e1d393c0e5a11bec964eb559dde6738ca58) to make sure I didn't mangle anything.

**Comment by RikkiGibson** [comment]
Looks like bootstrap build is failing only on Mac and Linux. Not sure why yet, but my code is on the stack trace, so I'm pretty sure it's related to the changes I made.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 10]
```diff
@@ -2825,6 +2826,103 @@ private void ReinferMethodAndVisitArguments(BoundCall node, TypeWithState receiv
...
+            var isObjectEqualsMethodOrOverride = method.GetLeastOverriddenMethod(accessingTypeOpt: null)
+                .Equals(compilation.GetSpecialTypeMember(SpecialMember.System_Object__Equals));
+            if (isObjectEqualsMethodOrOverride || isWellKnownEqualityMethodOrImplementation(method, WellKnownMember.System_IEquatable_T__Equals))
+            {
+                Debug.Assert(arguments.Length == 1);
+                learnFromEqualsMethodArguments(node.ReceiverOpt, receiverType, arguments[0], results[0].RValueType);
+                return;
+            }
+
+            bool isWellKnownEqualityMethodOrImplementation(MethodSymbol method, WellKnownMember wellKnownMember)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >bool [](start = 12, length = 4)

static?

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs [hunk 11]
```diff
@@ -1705,7 +1705,7 @@ private static Location GetInterfaceLocation(Symbol interfaceMember, TypeSymbol
                 snt = implementingType as SourceMemberContainerTypeSymbol;
             }
 
-            return snt?.GetImplementsLocation(@interface) ?? implementingType.Locations[0];
+            return snt?.GetImplementsLocation(@interface) ?? implementingType.Locations.FirstOrDefault();
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs]: >FirstOrDefault [](start = 88, length = 14)

potentially returning `null` Location here seems risky. Let's return `Location.None` instead in the worst case

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 12]
```diff
@@ -109782,6 +109782,120 @@ void verify(string source)
             }
         }
 
+        [Fact, WorkItem(35816, "https://github.com/dotnet/roslyn/issues/35816")]
+        public void StaticEqualsMethod_NonNullExpr_NotNullExpr_NoWarning()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: nit: `NonNull` versus `NotNull` inconsistency

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 13]
```diff
@@ -110493,5 +110724,217 @@ static void M1(C? c1)
...
+}";
+
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (6,21): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         object o2 = o1;
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "o1").WithLocation(6, 21));
+
+            Assert.True(comp.ShouldRunNullableWalker);
+        }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: nit: two empty lines below

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 14]
```diff
@@ -109735,6 +109735,996 @@ class Program
...
+        else
+        {
+            _ = s1.ToString();
+            _ = s2.ToString();
+        }
+    }
+}";
+            verify(source0);
+            verify(source0.Replace("object.Equals", "object.ReferenceEquals"));
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Nit: consider using a `Theory` instead, and having the argument for the test be `Equals` and `ReferenceEquals`

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Same suggestion for the test of the tests that do this style of multi-verification

---
In reply to: [298742762](https://github.com/dotnet/roslyn/pull/36722#discussion_r298742762) [](ancestors = 298742762)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 15]
```diff
@@ -109735,6 +109735,996 @@ class Program
...
+            _ = s.ToString();
+        }
+    }
+
+    static void M2(string s)
+    {
+        string? s2 = null;
+        if (object.Equals(s2, s))
+        {
+            _ = s.ToString();
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: This feels wrong. If this line of code was reached, we _will_ throw a NullReferenceException.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: I understand that this isn't a pure null test, so we can't learn anything from the comparison. But it's a good example of a scenario we could have caught if we differentiated between "This could have null in it" and "This _definitely_ has null in it", and vice versa for not null.

---
In reply to: [298755212](https://github.com/dotnet/roslyn/pull/36722#discussion_r298755212) [](ancestors = 298755212)

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Right. At the moment we have three categories ranging from highest confidence to lowest:
- a `null` literal or constant
- a not-null state
- a maybe-null state

In this test, we're comparing two expressions: one has a not-null state and the other has a maybe-null state. We're more confident in the not-null state, so we leave it alone.

File: src/Compilers/CSharp/Test/Symbol/Symbols/MissingSpecialMember.cs [hunk 16]
```diff
@@ -865,8 +866,8 @@ public void AllWellKnownTypesBeforeCSharp7()
                 Assert.True(type <= WellKnownType.CSharp7Sentinel);
             }
 
-            // There were 204 well-known types prior to CSharp7
-            Assert.Equal(204, (int)(WellKnownType.CSharp7Sentinel - WellKnownType.First));
+            // There were 205 well-known types prior to CSharp7
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Symbol/Symbols/MissingSpecialMember.cs]: Is this comment correct?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 17]
```diff
@@ -2825,6 +2826,105 @@ private void ReinferMethodAndVisitArguments(BoundCall node, TypeWithState receiv
...
+                foreach (var member in constructedType.GetMembers(WellKnownMemberNames.ObjectEquals))
+                {
+                    if (member.OriginalDefinition.Equals(wellKnownMethod))
+                    {
+                        constructedMethod = member;
+                        break;
+                    }
+                }
+
+                Debug.Assert(constructedMethod != null, "the original definition is present but the constructed method isn't present");
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Nit: empty line below

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs [hunk 18]
```diff
@@ -1705,7 +1705,7 @@ private static Location GetInterfaceLocation(Symbol interfaceMember, TypeSymbol
                 snt = implementingType as SourceMemberContainerTypeSymbol;
             }
 
-            return snt?.GetImplementsLocation(@interface) ?? implementingType.Locations[0];
+            return snt?.GetImplementsLocation(@interface) ?? implementingType.Locations.FirstOrNone();
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs]: >FirstOrNone [](start = 88, length = 11)

Is this Linq?

**Comment by 333fred** [comment]
Done review pass (commit 7). Some code cleanup comments.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs (hunk 18 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs]: Extension:
```C#
        internal static Location FirstOrNone(this ImmutableArray<Location> items)
        {
            return items.IsEmpty ? Location.None : items[0];
        }
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 15 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Will mark this resolved, let me know if there's action I should be taking related to this scenario (I do agree that it's questionable)

---
In reply to: [298757714](https://github.com/dotnet/roslyn/pull/36722#discussion_r298757714) [](ancestors = 298757714)

File: src/Compilers/CSharp/Test/Symbol/Symbols/MissingSpecialMember.cs (hunk 16 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/MissingSpecialMember.cs]: We should change this assertion to check `CSharp7Sentinel - First` instead. Then it remains true when adding new special types.
The point of `CSharp7Sentinel` is that our rules for finding well-known types changed a bit in C# 7 with regards to dealing with ambiguities:
```C#
                    // well-known types introduced before CSharp7 allow lookup ambiguity and report a warning
                    DiagnosticBag legacyWarnings = (type <= WellKnownType.CSharp7Sentinel) ? warnings : null;
```

---
In reply to: [298759023](https://github.com/dotnet/roslyn/pull/36722#discussion_r298759023) [](ancestors = 298759023)

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Symbol/Symbols/MissingSpecialMember.cs]: Should this new well-known type stay before the CSharp7Sentinel?

---
In reply to: [298769217](https://github.com/dotnet/roslyn/pull/36722#discussion_r298769217) [](ancestors = 298769217,298759023)

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Symbol/Symbols/MissingSpecialMember.cs]: after offline discussion it sounds like this should actually go after the sentinel so that we disallow multiple definitions of the type.

---
In reply to: [298770457](https://github.com/dotnet/roslyn/pull/36722#discussion_r298770457) [](ancestors = 298770457,298769217,298759023)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 15 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: No, there's nothing you can do, I'm just grouching.

---
In reply to: [298766107](https://github.com/dotnet/roslyn/pull/36722#discussion_r298766107) [](ancestors = 298766107,298757714)
